src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs

```diff
@@ -35,12 +35,31 @@ internal static class WebResponseHelper
             return headers;
         }
 
-        internal static string GetOutFilePath(HttpResponseMessage response, string _qualifiedOutFile)
-        {
-            // Get file name from last segment of Uri
-            string lastUriSegment = System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]);
+        internal static string GetOutFilePath(HttpResponseMessage response, string qualifiedOutFile)
+        {            
+            if (!Directory.Exists(qualifiedOutFile))
+            {
+                return qualifiedOutFile;
+            }
+
+            string contentDisposition = response.Content.Headers.ContentDisposition?.FileNameStar ?? response.Content.Headers.ContentDisposition?.FileName;
+
+            if (!string.IsNullOrEmpty(contentDisposition)) 
+            {
+                // Get file name from Content-Disposition header if present
+                return Path.Join(qualifiedOutFile, contentDisposition);
+            }
+
+            if (response.RequestMessage.RequestUri.PathAndQuery != "/")
+            {
+                string lastUriSegment = System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]);
+
+                // Get file name from last segment of Uri
+                return Path.Join(qualifiedOutFile, lastUriSegment);
+            }
 
-            return Directory.Exists(_qualifiedOutFile) ? Path.Join(_qualifiedOutFile, lastUriSegment) : _qualifiedOutFile;
+            // File name not found use sanitized Host name instead
+            return Path.Join(qualifiedOutFile, response.RequestMessage.RequestUri.Host.Replace('.', '_'));
```

Comment:
I want we explicitly confirm this fallback.
/cc @SteveL-MSFT @mklement0

Reply:
I compared with curl.
Curl simply fails autonaming if the url has no last segment-
![Cattura4](https://user-images.githubusercontent.com/105941898/227908982-79345ab7-e82e-4690-9111-76aaa78f924d.PNG)